### PR TITLE
Handle specification urls

### PR DESCRIPTION
**Pull request by obadrawi**
- configuration package handles the specification url and writes it to a temp file

### CODE REVIEW

File: modules/configuration/src/configuration.ts [hunk 1]
```diff
@@ -62,8 +72,19 @@ function resolveSpecs(config: Config) {
             })
         }
     }
+    return Promise.all(promises)
+}
+function getSpecFromUrl(url: string) {
+    return new Promise((resolve, reject) => {
+        request.get({
```
Comment (a-thaler) [inline modules/configuration/src/configuration.ts]: Overall I think you should use a callback approach for now to not have the "async" on the API interface.
Or really switch to it but without using an await in the code, you should be able to chain all the tasks without blocking.
Then you also need to change the API to be async and accept in the modules a Promise<Config>

File: modules/configuration/src/configuration.ts [hunk 2]
```diff
@@ -62,8 +72,19 @@ function resolveSpecs(config: Config) {
...
+            uri: url,
+            resolveWithFullResponse: true,
+            simple: false
+        }).then((response: any) => {
+            resolve(response.body)
```
Comment (a-thaler) [inline modules/configuration/src/configuration.ts]: I see no error handling!!! What if the URL is wrong or not reachable anymore? I want to have a nice validation error!

File: modules/openapi-mock/src/server/mock.ts [hunk 3]
```diff
@@ -6,7 +6,7 @@ import * as yaml from "js-yaml"
 import * as fs from "fs"
 import * as config from "@varkes/configuration"
 import { SwaggerMiddleware } from "swagger-express-middleware";
-
+import * as request from "request-promise";
```
[the remark this answers is not in the document]
Comment (obadrawi) [reply modules/openapi-mock/src/server/mock.ts]: removed

File: modules/openapi-mock/src/test/test.ts [hunk 4]
```diff
@@ -6,10 +6,13 @@ import * as  mock from "../server/app"
 import * as request from "supertest"
 import * as express from "express"
 import * as config from "@varkes/configuration"
-
+let resolvePath: any;
```
[the remark this answers is not in the document]
Comment (obadrawi) [reply modules/openapi-mock/src/test/test.ts]: change to type Config

File: modules/openapi-mock/src/server/server.ts [hunk 5]
```diff
@@ -15,7 +15,7 @@ var runAsync = async () => {
     }
 
     try {
-        let configuration = config.resolveFile(configPath, __dirname)
+        let configuration = await config.resolveFile(configPath, __dirname)
```
[the remark this answers is not in the document]
Comment (obadrawi) [reply modules/openapi-mock/src/server/server.ts]: as per our conversation, we will leave as is

File: modules/configuration/src/test/test.ts [hunk 6]
```diff
@@ -13,29 +13,33 @@ describe('config parser', () => {
     })
     it('fails with invalid JSON', (done) => {
         let input = ""
-        try {
-            configuration.resolve(input)
+        configuration.resolve(input).then((result) => {
```
[the remark this answers is not in the document]
Comment (obadrawi) [reply modules/configuration/src/test/test.ts]: as per our conversation, we will leave as is

File: modules/configuration/src/configuration.ts [hunk 7]
```diff
@@ -9,25 +9,26 @@ import { Config, API, Event } from "./types"
...
 const LOGGER = logger.logger("configuration")
-
+const URL_REGEX = /^((http|https):\/\/)[a-z0-9]+([\-\.]{1}[a-z0-9]+)*(\.[a-z]{2,5})*(:[0-9]{1,5})?(\/.*)?$/
 /**
  * Loads a config from a string
  * 
  * @param configText the string containing the configuration as text
  * @param location optional absolute path to the config in order to resolve referenced spec files
  */
-export function resolve(configText: string, location: string = ""): Config {
```
[the remark this answers is not in the document]
Comment (obadrawi) [reply modules/configuration/src/configuration.ts]: returned type

File: modules/configuration/src/configuration.ts (hunk 1 above)
Comment (obadrawi) [reply modules/configuration/src/configuration.ts]: as per our conversation, we will leave as is

File: modules/configuration/src/configuration.ts [hunk 8]
```diff
@@ -9,25 +9,26 @@ import { Config, API, Event } from "./types"
...
 const yaml = require("js-yaml");
 const pretty_yaml = require('json-to-pretty-yaml')
-
+import * as request from "request-promise";
+import * as tmp from "tmp"
 const OAUTH = "/authorizationserver/oauth/token";
 const METADATA = "/metadata";
 const LOGGER = logger.logger("configuration")
-
+const URL_REGEX = /^((http|https):\/\/)[a-z0-9]+([\-\.]{1}[a-z0-9]+)*(\.[a-z]{2,5})*(:[0-9]{1,5})?(\/.*)?$/
```
[the remark this answers is not in the document]
Comment (obadrawi) [reply modules/configuration/src/configuration.ts]: done..used the url library

File: modules/configuration/src/configuration.ts [hunk 9]
```diff
@@ -5,28 +5,29 @@ import path = require("path");
...
```
Comment (a-thaler) [inline modules/configuration/src/configuration.ts]: I still see the constant expression

File: modules/configuration/src/configuration.ts [hunk 10]
```diff
@@ -181,4 +204,4 @@ function validateEvents(config: Config) {
     }
 
     return errors
-}
```
[the remark this answers is not in the document]
Comment (a-thaler) [reply modules/configuration/src/configuration.ts]: ping

File: modules/configuration/src/configuration.ts [hunk 11]
```diff
@@ -62,8 +72,19 @@ function resolveSpecs(config: Config) {
             })
         }
     }
+    return Promise.all(promises)
+}
+function getSpecFromUrl(url: string) {
+    return new Promise((resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (obadrawi) [reply modules/configuration/src/configuration.ts]: removed promise

File: modules/configuration/src/configuration.ts [hunk 12]
```diff
@@ -37,23 +38,32 @@ export function resolve(configText: string, location: string = ""): Config {
...
-                api.specification = path.resolve(path.dirname(config.location), api.specification)
+            promises = promises.concat(config.apis.map(async (api: API) => {
+                if (!api.specification.match(URL_REGEX)) {
+                    api.specification = path.resolve(path.dirname(config.location), api.specification)
+                }
+                else {
+                    let body = await getSpecFromUrl(api.specification)
+                    var tmpobj = tmp.fileSync()
+                    fs.writeFileSync(tmpobj.name, body)
+                    api.specification = tmpobj.name
```
[the remark this answers is not in the document]
Comment (obadrawi) [reply modules/configuration/src/configuration.ts]: done

File: modules/openapi-mock/package.json [hunk 13]
```diff
@@ -17,13 +17,15 @@
...
   "dependencies": {
+    "@types/request-promise": "^4.1.44",
     "@varkes/configuration": "^0.8.0",
     "api-spec-converter": "~2.9.1",
     "body-parser": "^1.19.0",
     "express": "^4.17.1",
     "js-yaml": "^3.13.1",
     "json-to-pretty-yaml": "^1.2.2",
     "morgan": "^1.9.1",
+    "request-promise": "^4.2.5",
```
[the remark this answers is not in the document]
Comment (a-thaler) [reply modules/openapi-mock/package.json]: Why that request-promise dependency is needed?

File: modules/configuration/src/configuration.ts (hunk 2 above)
Comment (obadrawi) [reply modules/configuration/src/configuration.ts]: done

File: modules/configuration/src/configuration.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (obadrawi) [reply modules/configuration/src/configuration.ts]: we do support urls for events now

File: modules/configuration/src/configuration.ts [hunk 14]
```diff
@@ -37,23 +38,32 @@ export function resolve(configText: string, location: string = ""): Config {
...
 
-function resolveSpecs(config: Config) {
+async function resolveSpecs(config: Config) {
+    let promises: any = []
     if (config.location) {
         if (config.apis) {
-            config.apis.map((api: API) => {
-                api.specification = path.resolve(path.dirname(config.location), api.specification)
+            promises = promises.concat(config.apis.map(async (api: API) => {
+                if (!api.specification.match(URL_REGEX)) {
```
[the remark this answers is not in the document]
Comment (obadrawi) [reply modules/configuration/src/configuration.ts]: done

File: modules/openapi-mock/package.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (obadrawi) [reply modules/openapi-mock/package.json]: removed

**Comment by a-thaler** [comment]
Relates to https://github.com/kyma-incubator/varkes/issues/243

**Pull request merged**